lib/fuse_lowlevel.c

```diff
@@ -2835,6 +2838,8 @@ void fuse_session_process_buf_int(struct fuse_session *se,
 		do_write_buf(req, in->nodeid, inarg, buf);
 	else if (in->opcode == FUSE_NOTIFY_REPLY)
 		do_notify_reply(req, in->nodeid, inarg, buf);
+	else if (in->opcode == FUSE_INIT)
+		do_init(req, in->nodeid, inarg, buf->mem_size);
```

Comment:
You can't trust if this isn't called from an older libfuse version - at a minimum this needs to check a flag.
And with header/payload separation we can probably get entirely rid of the if conditions here - would be great if we could avoid another exception.

The issue I have in general is that with fuse-io-uring buffers are registered with FUSE_URING_REQ_FETCH and then pages are for privileged users persistently mapped. 

And then, what is if allocation at run time fail? Already bad enough that we (so far) dynamic threads, but now it also would do re-allocation.

Personally, I would really prefer to the decision about IO sizes at initialization time - which is already possible. We also might just set 1MB (or lower buffer sizes by default).

Reply:
Thanks for taking a look at this PR.

> You can't trust if this isn't called from an older libfuse version - at a minimum this needs to check a flag.

In the do_init() call, if the older libfuse version doesn't support `FUSE_MAX_PAGES`, then no reallocation for big write buffers happens.

> The issue I have in general is that with fuse-io-uring buffers are registered with FUSE_URING_REQ_FETCH and then pages are for privileged users persistently mapped.

For fuse-io-uring buffers, my understanding is that at the time buffers get registered with FUSE_URING_REQ_FETCH, the needed buf size would already be known (from the server's reply to the init call, which would go through through /dev/fuse before uring initialization is set up by the server).

 > And then, what is if allocation at run time fail? 

I think this would only fail for the out of memory case, which seems rare and could happen for any other allocation path at runtime too. 

> Already bad enough that we (so far) dynamic threads, but now it also would do re-allocation.

The re-allocation would only need to happen once. The main problem is that we need to allocate a buffer in order to set up / initialize the connection, but we only know if the server opts into bigger write buffers through their init callback. Unless instead, they mandate that they provide this information at mount time.

> Personally, I would really prefer to the decision about IO sizes at initialization time - which is already possible. We also might just set 1MB (or lower buffer sizes by default).

By 'initialization time', do you mean 'mount time'? I'm okay with this too - I just didn't like how clunky this interface gets ergonomically with having to opt into this at mount time. This is the approach I first submitted in https://github.com/libfuse/libfuse/pull/984, I'm happy to resurrect this and go with this approach if you prefer this approach instead.

Reply:
> Thanks for taking a look at this PR.
> 
> > You can't trust if this isn't called from an older libfuse version - at a minimum this needs to check a flag.
> 
> In the do_init() call, if the older libfuse version doesn't support `FUSE_MAX_PAGES`, then no reallocation for big write buffers happens.

There are also users in `fuse_session_receive_buf_int`, I just annotated it. And it is reading a field beyond old struct size (assuming you move the field at the end of the struct)  - I really think a flag would be needed if we go this way. 

> 
> > The issue I have in general is that with fuse-io-uring buffers are registered with FUSE_URING_REQ_FETCH and then pages are for privileged users persistently mapped.
> 
> For fuse-io-uring buffers, my understanding is that at the time buffers get registered with FUSE_URING_REQ_FETCH, the needed buf size would already be known (from the server's reply to the init call, which would go through through /dev/fuse before uring initialization is set up by the server).

Right now not, these /dev/fuse threads act on their own, we would basically need to wait for the reply to set things up. But then the reply might not come for a long time - if the /dev/fuse reading thread is waiting in fuse_dev_do_read(), because the provided buffer size is correct.  Also, fuse-io-uring is set up before libfuse even starts to read /dev/fuse, just the async nature of io-uring then typically cause FUSE_INIT to be complete before fuse-io-uring setup is complete.

With the current patch that uses UINT_MAX, I wouldn't know at all what to use as  fuse-io-uring buffer size.



> 
> > And then, what is if allocation at run time fail?
> 
> I think this would only fail for the out of memory case, which seems rare and could happen for any other allocation path at runtime too.

Other _large_ run time allocations are these dynamics threads. I won't look into any of that as I see it deprecated with io-uring.

> 
> > Already bad enough that we (so far) dynamic threads, but now it also would do re-allocation.
> 
> The re-allocation would only need to happen once. The main problem is that we need to allocate a buffer in order to set up / initialize the connection, but we only know if the server opts into bigger write buffers through their init callback. Unless instead, they mandate that they provide this information at mount time.
> 
> > Personally, I would really prefer to the decision about IO sizes at initialization time - which is already possible. We also might just set 1MB (or lower buffer sizes by default).
> 
> By 'initialization time', do you mean 'mount time'? I'm okay with this too - I just didn't like how clunky this interface gets ergonomically with having to opt into this at mount time. This is the approach I first submitted in #984, I'm happy to resurrect this and go with this approach if you prefer this approach instead.

That introduces se->extended_max_write`. why can't we use 'se->conn.max_write? 

Well, maybe I just move the initialization of io-uring post FUSE_INIT reply and we double ensure that new threads are not created as long as FUSE_INIT reply is not send. The FUSE_INIT reply would then still need to use the max size, that is currently read from /proc. I would also like to avoid the extra parameter for the do_init() function, if you don't want to store 'reallocable' into fuse_session, a local file variable should also do?

And alternative solution to reading /proc would involve kernel change, and fuse-client would send something like FUSE_INIT_PARAMS (or whatever it is called), with the parameters kernel has eventually chosen.

Reply:
> That introduces se->extended_max_write`. why can't we use 'se->conn.max_write?

We could use 'se->conn.max_write' to store the size, but the mount option will still need to be 'extended_max_write' since this would only need to be passed in for writes exceeding the default 1 MiB limit.

> I would also like to avoid the extra parameter for the do_init() function, if you don't want to store 'reallocable' into fuse_session, a local file variable should also do?

I'm fine with storing 'reallocable' as a field in 'struct fuse_session' instead of directly passing 'reallocable' as a parameter.